src/Test/StateMachine/Prototype.hs

```diff
@@ -450,10 +495,40 @@ liftGenFork gen pre next model = do
   (right,  _,      _) <- liftGen' gen pre next model' n
   return (Fork left prefix right)
 
+forkFilterPrecondition
+  :: HFoldable act
+  => Precondition model act
+  -> Transition model act
+  -> model Symbolic
+  -> Fork [Internal act]
+  -> Fork [Internal act]
+forkFilterPrecondition pre trans m (Fork l p r) =
+  Fork (snd $ filterPrecondition pre trans m' vars l)
+       p'
+       (snd $ filterPrecondition pre trans m' vars r)
+  where
+    ((m', vars), p') = filterPrecondition pre trans m Set.empty p
+
+shrinkPair
+  :: (a -> [a])
+  -> (b -> [b])
+  -> ((a, b) -> [(a, b)])
+shrinkPair shrinkA shrinkB (a, b) =
+     [ (a', b) | a' <- shrinkA a ]
+  ++ [ (a, b') | b' <- shrinkB b ]
+
 liftShrinkFork
-  :: (act v resp -> [act v resp])
+  :: HFoldable act
+  => (forall v resp. act v resp -> [act v resp])
+  -> Precondition model act
+  -> Transition model act
+  -> model Symbolic
   -> (Fork [Internal act] -> [Fork [Internal act]])
-liftShrinkFork _ _ = []
+liftShrinkFork oldShrink pre trans model (Fork l p r) =
+  map (forkFilterPrecondition pre trans model)
+  [ Fork l' p' r' | (p', (l', r')) <- shrinkPair shrinkSub (shrinkPair shrinkSub shrinkSub) (p, (l, r))]
```

Comment:
I'm not sure that is beneficial, I will leave it for now like it is for the similarity with `liftShrink`